src/vmm/src/devices/virtio/net/device.rs

```diff
@@ -556,29 +604,44 @@ impl Net {
                 self.rx_buffer
                     .iovec
                     .write_all_volatile_at(&self.rx_frame_buf[..vnet_hdr_len() + len], 0)?;
+
+                // SAFETY:
+                // * len is never 0
+                // * Maximum size of the packet is 65562 which fits into u32
+                #[allow(clippy::cast_possible_truncation)]
+                let len = unsafe { NonZeroU32::new_unchecked((vnet_hdr_len() + len) as u32) };
+
                 // SAFETY: This is safe:
                 // * We checked that `rx_buffer` includes at least one `DescriptorChain`
                 // * `rx_frame_buf` has size of `MAX_BUFFER_SIZE` and all `DescriptorChain` objects
                 //   are at least that big.
-                let dc = unsafe {
+                unsafe {
                     self.rx_buffer
-                        .mark_used((vnet_hdr_len() + len).try_into().unwrap())
-                };
-
-                return Ok(Some(dc));
+                        .mark_used(len.get(), &mut self.queues[RX_INDEX]);
+                }
+                return Ok(Some(len));
             }
         }
 
         // SAFETY: this is safe because we ensured that `self.rx_buffer` has at least one
         // DescriptorChain parsed in it.
         let len = unsafe { self.read_tap().map_err(NetError::IO) }?;
 
+        // SAFETY:
+        // * len is never 0
```

Comment:
I am not sure that this is true. `read` and `readv` might read less bytes than what we asked for, without returning an Error. Looking at the man pages, there doesn't seem to be any guarantees that these system calls cannot return 0 bytes.

In any case, how does the non-zero count guarantee help us, at all? What's the problem if `readv` returns 0?

Reply:
Added a comment. `readv` will return 0 if we pass it array of iovecs with 0 capacity. But we never do this.
The main thing with `NonZeroU32` was to keep existing `match` in the `process_rx`. I can swap it back to u32 if you and @roypat really want.

Reply:
`readv` does not return 0 only when we pass a buffer of 0. According to the `man read(2):

> It is not an error if this number is smaller than the number of bytes requested; this may happen for example because fewer bytes are actually available right now (maybe because we were close to end-of-file, or because we are reading from a pipe, or from a terminal)  __or because read() was interrupted by a signal__.

So, it seems to imply that reading might be interrupted without returning an `EINTR`.

But again, I don't think that `NonZeroU32` adds any value here.